ignite/contrib/handlers/ema_handler.py

```diff
@@ -0,0 +1,204 @@
+    def _unwrap_model(model: nn.Module) -> nn.Module:
+        if isinstance(model, (nn.parallel.DataParallel, nn.parallel.DistributedDataParallel)):
+            return model.module
+        else:
+            return model
+
+    def _get_momentum(self, curr_iter: int):
+        """Get current momentum, `curr_iter` should be 1-based. When `curr_iter = 1`, `momentum =
+        self.momentum_warmup`; when `curr_iter >= self.warmup_iters`, `momentum = self.momentum`"""
+        assert curr_iter >= 1
+        denominator = max(1, self.warmup_iters - 1)
+        momentum = self.momentum_warmup + (self.momentum - self.momentum_warmup) * (curr_iter - 1) / denominator
+        return min(self.momentum, momentum)
+
+    @torch.no_grad()
+    def _swap_params(self, engine: Engine, model: nn.Module) -> None:
+
+        model = self._unwrap_model(model)
+        for ema_v, model_v in zip(self.ema.state_dict().values(), model.state_dict().values()):
+            if self.device is not None:
+                model_v = model_v.to(device=self.device)
+            tmp = model_v.data.clone()
+            model_v.data.copy_(ema_v.data)
+            ema_v.data.copy_(tmp)
+
+    @torch.no_grad()
+    def _update(self, engine: Engine, model: nn.Module) -> None:
+
+        model = self._unwrap_model(model)
+        curr_iter = engine.state.iteration
+        momentum = self._get_momentum(curr_iter)
+
+        for ema_v, model_v in zip(self.ema.state_dict().values(), model.state_dict().values()):
+            if self.device is not None:
+                model_v = model_v.to(device=self.device)
+            ema_v.mul_(1 - momentum).add_(model_v.data, alpha=momentum)
+
+    def attach(self, engine: Engine, model: nn.Module) -> None:
+        """Attach the handler to engine.
+
+        Note:
+              If model is ``DataParallel`` or ``DistributedDataParallel``, the EMA smoothing will be applied to
+              ``model.module`` .
+
+        Args:
+            engine: Trainer to which the handler will be attached.
+            model: Model to which the EMA smoothing will be applied.
+
+        """
+        # swap the parameters at the end of each epoch. If using Events.EPOCH_COMPLETED here, the handler of swap
+        # parameters might be called after validation handler.
+        # TODO Find a better hack to ensure that handler of swapping parameter is called before validation.
+        num_handlers_epoch_completed = engine._event_handlers[Events.EPOCH_COMPLETED]
+        if len(num_handlers_epoch_completed) > 0:
+            warnings.warn(
+                f"engine already has {num_handlers_epoch_completed} handlers at Events.EPOCH_COMPLETED, "
+                f"please manually check if the EMA handler is attached to the engine before other "
+                f"handlers"
+            )
+        engine.add_event_handler(Events.EPOCH_COMPLETED, self._swap_params, model)
+        engine.add_event_handler(Events.EPOCH_STARTED, self._swap_params, model)
```

Comment:
Could you please explain why we need to swap the params every epoch started and completed ? 
Wouldn't be better to expose ema model somehow to the user and use it for validation directly ?

Reply:
Please consider the case when users use DDP training. If we let the users to use the EMA model for validation, they need to initialize an extra DDP model for multi-gpu validation, right? Will it increase unnecessary communication cost? Please correct me if I am wrong.

Reply:
Actually, DDP wrapping is only needed if model params require grad (see https://pytorch.org/docs/master/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel). We can consider EMA model as one with params without grads and thus we do not need to wrap it with DDP...

Reply:
@vfdev-5 Thanks for the explanation. I have removed the functions for swapping parameters, and expose the ema model with a function `get_ema_model`.